File: anna/spec.js
// anna/spec.js
// ─────────────────────────────────────────────────────────────────────────────
// Carrier Spec compiler.
//
// Anna's matching model is two-layer (see matcher.js):
//   1. HARD GATES  — pass/fail disqualifiers. Any failure => INELIGIBLE.
//   2. SOFT WEIGHTS — preferences used only to RANK carriers a driver passes.
//
// This module turns a carrier's stored hiring requirements (QuickHire's nested
// `requirements` shape: { cdl, mvr, psp, insurance, ... }) into a reusable,
// versioned CarrierSpec made of declarative gates + weights. Gates are DATA, not
// code, so new rule types are additive and the result is auditable/testable.
// ─────────────────────────────────────────────────────────────────────────────

/**
 * @typedef {Object} Gate
 * @property {string} id        Stable identifier (e.g. "mvr.maxDUI").
 * @property {string} label     Human-readable label for summaries.
 * @property {string} field     Dot-path into the DriverProfile (e.g. "mvr.dui").
 * @property {string} op        One of OPS keys.
 * @property {*}      value     Comparison value from the carrier requirement.
 * @property {string} category  CDL | MVR | PSP | Insurance | Eligibility.
 */

/**
 * @typedef {Object} CarrierSpec
 * @property {string} carrierId
 * @property {string} carrierName
 * @property {number} version             Bumped whenever requirements change.
 * @property {Gate[]} hardGates
 * @property {Object<string,number>} softWeights  factorId -> weight (0..1).
 * @property {Object} raw                 Original requirements (kept for the LLM / audit).
 */

// Supported gate operators. Each returns true when the driver PASSES the gate.
// `d` = driver value (already normalized), `v` = carrier requirement value.
const CDL_RANK = { A: 3, B: 2, C: 1 };
const num = (x) => (Number.isFinite(Number(x)) ? Number(x) : null);

export const OPS = {
  gte: (d, v) => num(d) >= num(v),
  lte: (d, v) => num(d) <= num(v),
  eq: (d, v) => d === v,
  cdlRankGte: (d, v) => (CDL_RANK[String(d || '').toUpperCase()] || 0) >= (CDL_RANK[String(v || '').toUpperCase()] || 0),
  includesAll: (d, v) => Array.isArray(v) && v.every((x) => (Array.isArray(d) ? d : []).includes(x)),
  notOwnerOperator: (d /* driver.cdl.type */) => d !== 'owner-operator',
};

// Which ops need a non-null driver value to be evaluable. If the driver value is
// missing for one of these, the gate is UNKNOWN (needs data) rather than FAIL.
const NUMERIC_OPS = new Set(['gte', 'lte']);

/**
 * Resolve a dot-path from an object (e.g. "mvr.dui").
 * @returns {*} the value, or undefined if any segment is missing.
 */
export function getPath(obj, path) {
  return String(path).split('.').reduce((o, k) => (o == null ? undefined : o[k]), obj);
}

/**
 * Compile a carrier record into a CarrierSpec.
 * Accepts the QuickHire carrier object ({ id, name, requirements }) OR a bare
 * requirements object. Unknown / blank requirement fields are simply skipped, so
 * a carrier that only specifies a few rules just gets a few gates.
 *
 * @param {Object} carrier
 * @returns {CarrierSpec}
 */
export function compileSpec(carrier = {}) {
  const reqs = carrier.requirements || carrier || {};
  const gates = [];
  const push = (g) => { if (g.value != null && g.value !== '' && !(Array.isArray(g.value) && g.value.length === 0)) gates.push(g); };

  // ── Eligibility (general FMCSA-style gates) ──
  const elig = reqs.eligibility || {};
  push({ id: 'eligibility.minAge', label: `Minimum age ${elig.minAge}`, field: 'age', op: 'gte', value: num(elig.minAge), category: 'Eligibility' });

  // ── CDL ──
  const cdl = reqs.cdl || {};
  push({ id: 'cdl.class', label: `CDL Class ${cdl.class} or higher`, field: 'cdl.class', op: 'cdlRankGte', value: cdl.class, category: 'CDL' });
  push({ id: 'cdl.endorsements', label: `Endorsements: ${(cdl.endorsements || []).join(', ')}`, field: 'cdl.endorsements', op: 'includesAll', value: cdl.endorsements, category: 'CDL' });
  push({ id: 'cdl.minExperienceYears', label: `Min ${cdl.minExperienceYears} yr(s) experience`, field: 'cdl.experienceYears', op: 'gte', value: num(cdl.minExperienceYears), category: 'CDL' });
  push({ id: 'cdl.minValidityDays', label: `CDL valid ${cdl.minValidityDays}+ days`, field: 'cdl.expiresInDays', op: 'gte', value: num(cdl.minValidityDays), category: 'CDL' });
  if (cdl.allowOwnerOperator === false) push({ id: 'cdl.noOwnerOperator', label: 'No owner-operators', field: 'cdl.type', op: 'notOwnerOperator', value: false, category: 'CDL' });

  // ── MVR ──
  const mvr = reqs.mvr || {};
  push({ id: 'mvr.maxMovingViolations', label: `Max ${mvr.maxMovingViolations} moving violation(s)`, field: 'mvr.movingViolations', op: 'lte', value: num(mvr.maxMovingViolations), category: 'MVR' });
  push({ id: 'mvr.maxAccidents', label: `Max ${mvr.maxAccidents} accident(s)`, field: 'mvr.accidents', op: 'lte', value: num(mvr.maxAccidents), category: 'MVR' });
  push({ id: 'mvr.maxDUI', label: `Max ${mvr.maxDUI} DUI/DWI`, field: 'mvr.dui', op: 'lte', value: num(mvr.maxDUI), category: 'MVR' });

  // ── PSP ──
  const psp = reqs.psp || {};
  push({ id: 'psp.maxCrashes', label: `Max ${psp.maxCrashes} PSP crash(es)`, field: 'psp.crashes', op: 'lte', value: num(psp.maxCrashes), category: 'PSP' });
  push({ id: 'psp.maxOOSInspections', label: `Max ${psp.maxOOSInspections} OOS inspection(s)`, field: 'psp.oosInspections', op: 'lte', value: num(psp.maxOOSInspections), category: 'PSP' });

  // ── Insurance ──
  const ins = reqs.insurance || {};
  push({ id: 'insurance.minAutoLiability', label: `Min $${num(ins.minAutoLiability)?.toLocaleString?.() || ins.minAutoLiability} auto liability`, field: 'insurance.autoLiability', op: 'gte', value: num(ins.minAutoLiability), category: 'Insurance' });
  if (ins.cargoRequired) push({ id: 'insurance.minCargo', label: `Min $${num(ins.minCargo)?.toLocaleString?.() || ins.minCargo} cargo`, field: 'insurance.cargo', op: 'gte', value: num(ins.minCargo), category: 'Insurance' });

  return {
    carrierId: carrier.id || reqs.id || null,
    carrierName: carrier.name || reqs.name || 'Unknown carrier',
    version: carrier.specVersion || carrier.version || 1,
    hardGates: gates,
    softWeights: normalizeWeights(reqs.softWeights),
    raw: reqs,
  };
}

// Default soft-ranking factors. Carriers may override via reqs.softWeights.
const DEFAULT_WEIGHTS = {
  experienceMargin: 0.35, // experience above the minimum
  cleanRecordMargin: 0.30, // violations/accidents/DUI well under the caps
  endorsementsMatch: 0.20, // holds the requested endorsements (and extras)
  cdlValidityMargin: 0.15, // CDL valid comfortably beyond the minimum
};

function normalizeWeights(overrides) {
  const w = { ...DEFAULT_WEIGHTS, ...(overrides || {}) };
  const total = Object.values(w).reduce((s, x) => s + (Number(x) || 0), 0) || 1;
  // Re-normalize so weights always sum to 1 (keeps fit scores comparable across carriers).
  return Object.fromEntries(Object.entries(w).map(([k, v]) => [k, (Number(v) || 0) / total]));
}

export { NUMERIC_OPS };

File: backend/app/anna/spec.py
"""Carrier Spec compiler — turns requirements into hard gates + soft weights."""

CDL_RANK = {"A": 3, "B": 2, "C": 1}


def _num(x):
    try:
        n = float(x)
        return n if n == n else None
    except (TypeError, ValueError):
        return None


def _includes_all(d, v):
    if not isinstance(v, list):
        return False
    dl = d if isinstance(d, list) else []
    return all(x in dl for x in v)


OPS = {
    "gte": lambda d, v: (_num(d) is not None and _num(v) is not None and _num(d) >= _num(v)),
    "lte": lambda d, v: (_num(d) is not None and _num(v) is not None and _num(d) <= _num(v)),
    "eq": lambda d, v: d == v,
    "cdlRankGte": lambda d, v: CDL_RANK.get(str(d or "").upper(), 0) >= CDL_RANK.get(str(v or "").upper(), 0),
    "includesAll": _includes_all,
    "notOwnerOperator": lambda d, v=None: d != "owner-operator",
}

NUMERIC_OPS = {"gte", "lte"}


def get_path(obj, path):
    cur = obj
    for k in str(path).split("."):
        if cur is None:
            return None
        cur = cur.get(k) if isinstance(cur, dict) else None
    return cur


def _fmt(n):
    try:
        return f"{int(n):,}"
    except (TypeError, ValueError):
        return n


def compile_spec(carrier: dict | None = None) -> dict:
    carrier = carrier or {}
    reqs = carrier.get("requirements") or carrier or {}
    gates = []

    def push(g):
        v = g["value"]
        if v is not None and v != "" and not (isinstance(v, list) and len(v) == 0):
            gates.append(g)

    elig = reqs.get("eligibility") or {}
    push({"id": "eligibility.minAge", "label": f"Minimum age {elig.get('minAge')}", "field": "age", "op": "gte", "value": _num(elig.get("minAge")), "category": "Eligibility"})

    cdl = reqs.get("cdl") or {}
    push({"id": "cdl.class", "label": f"CDL Class {cdl.get('class')} or higher", "field": "cdl.class", "op": "cdlRankGte", "value": cdl.get("class"), "category": "CDL"})
    push({"id": "cdl.endorsements", "label": f"Endorsements: {', '.join(cdl.get('endorsements') or [])}", "field": "cdl.endorsements", "op": "includesAll", "value": cdl.get("endorsements"), "category": "CDL"})
    push({"id": "cdl.minExperienceYears", "label": f"Min {cdl.get('minExperienceYears')} yr(s) experience", "field": "cdl.experienceYears", "op": "gte", "value": _num(cdl.get("minExperienceYears")), "category": "CDL"})
    push({"id": "cdl.minValidityDays", "label": f"CDL valid {cdl.get('minValidityDays')}+ days", "field": "cdl.expiresInDays", "op": "gte", "value": _num(cdl.get("minValidityDays")), "category": "CDL"})
    if cdl.get("allowOwnerOperator") is False:
        push({"id": "cdl.noOwnerOperator", "label": "No owner-operators", "field": "cdl.type", "op": "notOwnerOperator", "value": False, "category": "CDL"})

    mvr = reqs.get("mvr") or {}
    push({"id": "mvr.maxMovingViolations", "label": f"Max {mvr.get('maxMovingViolations')} moving violation(s)", "field": "mvr.movingViolations", "op": "lte", "value": _num(mvr.get("maxMovingViolations")), "category": "MVR"})
    push({"id": "mvr.maxAccidents", "label": f"Max {mvr.get('maxAccidents')} accident(s)", "field": "mvr.accidents", "op": "lte", "value": _num(mvr.get("maxAccidents")), "category": "MVR"})
    push({"id": "mvr.maxDUI", "label": f"Max {mvr.get('maxDUI')} DUI/DWI", "field": "mvr.dui", "op": "lte", "value": _num(mvr.get("maxDUI")), "category": "MVR"})

    psp = reqs.get("psp") or {}
    push({"id": "psp.maxCrashes", "label": f"Max {psp.get('maxCrashes')} PSP crash(es)", "field": "psp.crashes", "op": "lte", "value": _num(psp.get("maxCrashes")), "category": "PSP"})
    push({"id": "psp.maxOOSInspections", "label": f"Max {psp.get('maxOOSInspections')} OOS inspection(s)", "field": "psp.oosInspections", "op": "lte", "value": _num(psp.get("maxOOSInspections")), "category": "PSP"})

    ins = reqs.get("insurance") or {}
    minliab = _num(ins.get("minAutoLiability"))
    push({"id": "insurance.minAutoLiability", "label": f"Min ${_fmt(minliab) if minliab is not None else ins.get('minAutoLiability')} auto liability", "field": "insurance.autoLiability", "op": "gte", "value": minliab, "category": "Insurance"})
    if ins.get("cargoRequired"):
        mincargo = _num(ins.get("minCargo"))
        push({"id": "insurance.minCargo", "label": f"Min ${_fmt(mincargo) if mincargo is not None else ins.get('minCargo')} cargo", "field": "insurance.cargo", "op": "gte", "value": mincargo, "category": "Insurance"})

    return {
        "carrierId": carrier.get("id") or reqs.get("id"),
        "carrierName": carrier.get("name") or reqs.get("name") or "Unknown carrier",
        "version": carrier.get("specVersion") or carrier.get("version") or 1,
        "hardGates": gates,
        "softWeights": _normalize_weights(reqs.get("softWeights")),
        "raw": reqs,
    }


DEFAULT_WEIGHTS = {
    "experienceMargin": 0.35,
    "cleanRecordMargin": 0.30,
    "endorsementsMatch": 0.20,
    "cdlValidityMargin": 0.15,
}


def _normalize_weights(overrides):
    w = {**DEFAULT_WEIGHTS, **(overrides or {})}
    total = sum((_num(x) or 0) for x in w.values()) or 1
    return {k: (_num(v) or 0) / total for k, v in w.items()}
